_posts/2019-09-25-scatter-gather-pt1.md

```diff
@@ -0,0 +1,99 @@
+---
+layout: post
+title: Scatter-Gather in NServiceBus, part 1
+---
+
+This article series is about showing how to implement the Scatter-Gather pattern in NServiceBus. There are some common mistakes and hidden traps users can easily fall into, leading to bad performance and other issues.
+
+Scatter-Gather is one of the patterns mentioned in the [Enterprise Integration Patterns book](https://www.amazon.com/Enterprise-Integration-Patterns-Designing-Deploying/dp/0321200683). [It is described as](https://www.enterpriseintegrationpatterns.com/patterns/messaging/BroadcastAggregate.html):
+
+> The Scatter-Gather routes a request message to the a number of recipients. It then uses an Aggregator to collect the responses and distill them into a single response message.
+
+The pattern consists of two parts: Broadcasting a message to multiple recipients and then using the [Aggregator pattern](https://www.enterpriseintegrationpatterns.com/patterns/messaging/Aggregator.html) to gather responses.
+
+
+### Broadcast
+
+There are multiple ways to "broadcast" a message to multiple recipients. Typically, broadcasting is associated with [Publish-Subscribe](https://www.enterpriseintegrationpatterns.com/patterns/messaging/PublishSubscribeChannel.html) which is implemented using the `Publish` API in NServiceBus. However, there are many different patterns to send messages to multiple recipients and it depends on the type of the message, the recipients and the form of coupling between sender and receivers to find the best way to "broadcast".
+In this blog post, I'm going to use the [Splitter pattern](https://www.enterpriseintegrationpatterns.com/patterns/messaging/Sequencer.html): An incoming message is split into multiple smaller messages which can then be processed independently (and parallel) from each other. The result of each "sub-message" is then aggregated back to a single result again. This pattern is implemented using regular `Send` operations in NServiceBus. Strictly speaking, this might not match 100% with the definition of the Scatter-Gather pattern as we're sending multiple messages of the same type (with different values) but for the remainer of this blog post, it doesn't make a big difference whether you publish an event or send commands.
```

Comment:
```suggestion
In this blog post, I'm going to use the [Splitter pattern](https://www.enterpriseintegrationpatterns.com/patterns/messaging/Sequencer.html): An incoming message is split into multiple smaller messages which can then be processed independently (and in parallel) from each other. The result of each "sub-message" is then aggregated back to a single result again. This pattern is implemented using regular `Send` operations in NServiceBus. Strictly speaking, this might not match 100% with the definition of the Scatter-Gather pattern as we're sending multiple messages of the same type (with different values) but for the remainer of this blog post, it doesn't make a big difference whether you publish an event or send commands.
```